qkeras/quantizers.py

```diff
@@ -439,9 +440,537 @@ def trainable_variables(self):
+
+  @use_stochastic_rounding.setter
+  def use_stochastic_rounding(self, use_stochastic_rounding):
+    self._use_stochastic_rounding = use_stochastic_rounding
+
+  @property
+  def scale_axis(self):
+    return self._scale_axis
+
+  @scale_axis.setter
+  def scale_axis(self, scale_axis):
+    self._scale_axis = scale_axis
+
+  @property
+  def alpha(self):
+    return self._alpha
+
+  @alpha.setter
+  def alpha(self, alpha):
+    """Set alpha and check if it isvalid."""
+
+    self._alpha = alpha
+
+    # Error checking
+    if isinstance(alpha, six.string_types):
+      # Check the quantizer has been given a valid alpha string
+      if not alpha in self.ALPHA_OPTIONS:
+        raise ValueError(
+            f"Invalid alpha '{alpha}' for auto alpha computation. "
+            f"Must be one of {self.ALPHA_OPTIONS}")
+    elif alpha is not None: # alpha is a tensor
+      try:
+        # any allowable array type can be cast as a numpy array
+        np.array(alpha)
+      except TypeError:
+        raise TypeError(
+            f"alpha must be, a string, an array, or None, not {type(alpha)}")
+
+    if self._initialized:
+      self._set_default_quantization_scale()
+
+  @property
+  def scale(self):
+    return self.quantization_scale / self.data_type_scale
+  
+  @property
+  def quantization_scale(self):
+    return self._quantization_scale
+  
+  @quantization_scale.setter
+  def quantization_scale(self, quantization_scale):
+    self._quantization_scale = quantization_scale
+    
+  @property
+  def data_type_scale(self):
+    """Quantization scale for the data type"""
+    # integer is sometimes cast as int32, so cast to float32 to avoid errors
+    integer = tf.cast(self.integer, tf.float32)
+    return K.pow(2.0, integer - self.bits + self.keep_negative)
+
+  @property
+  def auto_alpha(self):
+    """Returns true if using a data-dependent alpha"""
+
+    return isinstance(self.alpha, six.string_types)
+
+  @property
+  def use_sign_function(self):
+    """Return true if using sign function for quantization"""
+
+    return (self.bits == 1.0) and self.keep_negative
+
+  @property
+  def clip_bounds(self):
+    """Get bounds of clip range"""
+
+    if self.use_sign_function:
+      clip_min = K.cast_to_floatx(-0.5)
+      clip_max = K.cast_to_floatx(0.5)
+    else:
+      unsigned_bits_po2 = K.pow(2.0, self.bits - self.keep_negative)
+      # if symmetric, clip_min is negative of clip_max. Otherwise clip_min is
+      # lowered by 1, giving us one more representable number
+      clip_min = self.keep_negative * (-unsigned_bits_po2 + self.symmetric)
+      clip_max = unsigned_bits_po2 - K.cast_to_floatx(1.0)
+
+    return clip_min, clip_max
+  
+  def _set_default_quantization_scale(self):
+    """Calculate and set quantization_scale default"""
+    assert (
+        self._initialized
+    ), "Must initialize before calling _set_default_quantization_scale"
+
+
+    err_msg = (f"Bit count {self.bits} must exceed "
```

Comment:
Why not put the error message inside the ValueError?

Reply:
Just to save space and not go over multiple lines. I will move it into the if statement